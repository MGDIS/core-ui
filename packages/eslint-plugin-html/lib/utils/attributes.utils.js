module.exports = {
  /**
   *  Check if node had an attribute
   *
   * @param {object} node node to check
   * @param {string} attribute attribute we are looking for
   * @returns {object} matching attribute
   */
  attribute(node, attribute) {
    return node.attrs?.find(attr => attr.name === attribute);
  },
  /**
   * Check if node has a specific class name
   *
   * @param {object} node
   * @param {string} className
   * @returns {Boolean} has class name
   */
  hasClass(node, className) {
    const classesAttribute = module.exports.attribute(node, 'class')?.value.split(' ');
    return classesAttribute && classesAttribute.some(classAttribute => classAttribute === className);
  },
};
