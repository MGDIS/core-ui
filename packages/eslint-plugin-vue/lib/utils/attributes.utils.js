module.exports = {
  /**
   *  Check if node had an attribute
   *
   * @param {object} node node to check
   * @param {string} attribute attribute we are looking for
   * @returns {object} matching attribute
   */
  attribute(node, attribute) {
    return node.startTag?.attributes?.find(attr => {
      return attr.directive ? attr.key.argument?.name === attribute : attr.key.name === attribute;
    });
  },
  /**
   * Check if node has a specific class name
   *
   * @param {object} node
   * @param {string} className
   * @returns {Boolean} has class name
   */
  hasClass(node, className) {
    const classesAttribute = module.exports.getAttributeValue(module.exports.attribute(node, 'class'))?.split(' ');
    return classesAttribute && classesAttribute.some(classAttribute => classAttribute === className);
  },
  /**
   * Get attribute value
   * If the attribute is a directive, the path to get the value is different
   *
   * @param {object} attribute
   * @returns {string} attribute value
   */
  getAttributeValue(attribute) {
    let attributeValue;
    if (attribute?.directive) {
      // directive attributes (e.g. :id="withProps")
      attributeValue = module.exports.getValue(attribute.value?.expression);
      if (attribute?.key?.argument?.name == 'id') {
        // If the attribute is an ID, we remove spaces, done for the ConditionalExpression case
        attributeValue = attributeValue.split(' ').join('-');
      }
    } else {
      attributeValue = attribute?.value?.value;
    }
    return attributeValue ?? '';
  },
  /**
   * Get expression value based on type
   *
   * @param {object} expression
   * @returns {string} expression value
   */
  getValue(expression) {
    switch (expression?.type) {
      case 'Literal': // Used for values (e.g. :id="'kebab-case'")
        return expression.value;
      case 'Identifier': // Used for variables (e.g. :id="withProps")
        return expression.name.toLowerCase();
      case 'CallExpression': // Used for function calls (e.g. :id="method(blu)")
        return 'call-expression'; // No need to reconstruct
      case 'TemplateLiteral': // Used for template literals (e.g. :id="\`with-\${props}\`")
        return module.exports.reconstructTemplateLiteral(expression);
      case 'ConditionalExpression': // Used for ternary operators (e.g. :id="using ? 'ternary' : 'operator'")
        return `${module.exports.getValue(expression.consequent)} ${module.exports.getValue(expression.alternate)}`;
      case 'ObjectExpression': // Used for object expressions (e.g. :id="{ 'blu': blu }")
        return expression.properties.map(property => module.exports.getValue(property.key)).join(' ');
      case 'ArrayExpression': // Used for array expressions (e.g. :id="['rich-text-input__input-container', { 'rich-text-input__input-container--readonly': readonly }]")
        return expression.elements.map(element => module.exports.getValue(element)).join(' ');
      case 'BinaryExpression': // Used for binary expressions (e.g. :id="blu + 'bli'")
        return `${module.exports.getValue(expression.left)}${module.exports.getValue(expression.right)}`;
      case 'MemberExpression': // Used for member expressions (e.g. :id="blu.bli")
        return 'member-expression'; // No need to reconstruct
      default:
        return '';
    }
  },
  /**
   * Reconstruct a template literal from its expression
   *
   * @param {object} expression
   * @returns {string} reconstructed template literal
   */
  reconstructTemplateLiteral(expression) {
    let templateLiteral = '';
    for (let i = 0; i < expression.quasis.length; i++) {
      const quasi = expression.quasis[i];
      templateLiteral += quasi.value.cooked ?? '';
      if (i < expression.expressions.length) {
        templateLiteral += module.exports.getValue(expression.expressions[i]);
      }
    }
    return templateLiteral;
  },
};
