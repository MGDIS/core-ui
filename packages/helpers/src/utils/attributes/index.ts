export * from './class-list';
export * from './id';
export * from './to-kebab-case';
