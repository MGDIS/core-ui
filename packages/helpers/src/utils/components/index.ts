export * from './components';
export type { CursorType } from './components.conf';
