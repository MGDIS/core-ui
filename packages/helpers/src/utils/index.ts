export * from './attributes';
export * from './components';
export * from './date';
export * from './elements';
export * from './number';
export * from './object';
export * from './string';
export * from './windows';
