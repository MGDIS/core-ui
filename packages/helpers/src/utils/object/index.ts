export * from './object';
export * from './object.conf';
