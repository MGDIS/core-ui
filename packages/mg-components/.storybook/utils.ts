import { renderVdom } from '@stencil/core/internal/client';

export const filterArgs = (args, defaultValues?) => {
  const filteredArgs: typeof args = {};
  for (const k in args) {
    if (!k.startsWith('slot')) {
      const arg = args[k];
      // Change camelCase k to kebab-case
      const key = k.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase();
      if (!defaultValues || !Object.keys(defaultValues).includes(k) || defaultValues[k] !== arg) {
        filteredArgs[key] = arg;
      }
    }
  }
  return filteredArgs;
};

export const stencilWrapper = (storyFn, context): Element => {
  const host = document.getElementById('storybook-root');
  renderVdom(
    {
      $ancestorComponent$: undefined,
      $flags$: 0,
      $modeName$: undefined,
      $cmpMeta$: {
        $flags$: 0,
        $tagName$: host.tagName,
      },
      $hostElement$: host,
    },
    storyFn(context),
  );
  return host.children[host.children.length - 1];
};

const renderAttribute = (element, name, value): void => {
  if ([null, undefined, '', false].includes(value) || ['innerHTML', 'style'].includes(name)) return;

  element.setAttribute(name, typeof value !== 'object' ? value : `/!\\ Object props are not rendered in the code example`);
};

const renderElement = (parentNode, tagName, attributes, children, text): void => {
  if (text) {
    parentNode.innerText = text;
    return;
  } else if (!tagName) return;
  const element = document.createElement(tagName);
  Object.keys(attributes || {}).forEach(attr => {
    renderAttribute(element, attr, attributes[attr]);
  });

  children?.forEach(child => {
    renderElement(element, child.$tag$, child.$attrs$, child.$children$, child.$text$);
  });

  if (attributes?.innerHTML) element.innerHTML = attributes.innerHTML;

  parentNode.appendChild(element);
};

export const getStoryHTML = ({ $tag$, $attrs$, $children$, $text$ }): string => {
  const host = document.createElement('div');

  renderElement(host, $tag$, $attrs$, $children$, $text$);

  return host.innerHTML;
};
