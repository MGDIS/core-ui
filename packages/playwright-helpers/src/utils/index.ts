/**
 * Render attributes from props objects
 * @param args - argument to render as string. ex: `{status: 'visible'}`
 * @returns formated inline attributed. ex: 'status="visible"'
 * @example
 * ```ts
 * import { renderAttributes } from '@mgdis/playwright-helpers';
 *
 * const attributes = renderAttributes({ status: 'visible', color: 'red' });
 * console.log(attributes); // Output: 'status="visible" color="red"'
 * ```
 */
export const renderAttributes = (args: Record<string, unknown>): string =>
  (args !== null &&
    typeof args === 'object' &&
    Object.keys(args)
      .filter(key => ![null, undefined, false].includes((args as Record<string, never>)[key]) && !['object', 'function'].includes(typeof (args as Record<string, never>)[key]))
      .map(key => `${key.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase()}="${(args as Record<string, never>)[key]}"`)
      .join(' ')) ||
  '';

/**
 * Render properties from props objects.
 * Insert return value in <script></script> element
 * @param args - argument to render as script. ex: `{status: 'visible'}`
 * @param selector - querySelector get targetted element and bind properties on it
 * @returns stringified properties script
 * @example
 * ```ts
 * import { renderProperties } from '@mgdis/playwright-helpers';
 *
 * const propertiesScript = renderProperties({ status: 'visible', color: 'red' }, '.targetElement');
 * console.log(propertiesScript);
 * // Output: 'document.querySelector(".targetElement").status="visible"; document.querySelector(".targetElement").color="red";'
 * ```
 */
export const renderProperties = (args: Record<string, unknown>, selector: string): string => {
  // has generated id selector starting with "#" create charcters (ex: '#0j3xzw4w7m') make test crashed sometime with "querySelector"
  // we use "getElementById" document query
  const query: 'querySelector' | 'getElementById' = selector.startsWith('#') ? 'getElementById' : 'querySelector';
  if (query === 'getElementById') selector = selector.replace('#', '');

  return args !== null && typeof args === 'object' && Object.keys(args).length > 0
    ? `
  ${
    Object.keys(args)
      .filter(key => ['object', 'function'].includes(typeof (args as Record<string, never>)[key]))
      .map(
        key =>
          `document.${query}('${selector}').${key}=${JSON.stringify((args as Record<string, never>)[key], (_key, val) => (typeof val === 'function' ? `<fn>${val}</fn>` : val))}`,
      ) // stringify json AND keep function values
      .join(';') // create string
      .split('"<fn>') // remove fn start decorator
      .join('')
      .split('</fn>"')
      .join('') // remove fn end decorator
  }`
    : '';
};
