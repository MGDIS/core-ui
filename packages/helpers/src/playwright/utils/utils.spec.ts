import { describe, expect, test } from 'vitest';
import { renderAttributes, renderProperties } from './utils';

describe('utils', () => {
  describe('renderAttributes', () => {
    test.each([
      undefined,
      null,
      {},
      { name: 'batman' },
      { name: '<div class="blu">blu</div>' },
      { name: 'batman', user: 'bruce', active: true, id: 1, object: {}, array: [], function: () => {} },
      { emptyString: '' },
    ])('Should render attributes', args => {
      const res = renderAttributes(args as Record<string, unknown>);
      expect(res).toMatchSnapshot();
    });
  });

  describe('renderProperties', () => {
    test('Should render properties, case id selector', () => {
      const res = renderProperties({ object: {} }, '#id');
      expect(res).toEqual("\n  document.getElementById('id').object={}");
    });
    test.each([undefined, null, {}, { name: 'batman' }])('Should not render attributes as properties', args => {
      const res = renderProperties(args as Record<string, unknown>, 'div');
      expect(res).toMatchSnapshot();
    });
    test.each([undefined, () => {}].map(fn => ({ name: 'batman', user: 'bruce', active: true, id: 1, object: {}, array: [], fn })))(
      'Should render properties, case id selector',
      args => {
        const res = renderProperties(args, 'div');
        expect(res).toMatchSnapshot();
      },
    );
  });
});
