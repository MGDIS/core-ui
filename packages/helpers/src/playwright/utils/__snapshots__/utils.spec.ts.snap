// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils > renderAttributes > Should render attributes 1`] = `""`;

exports[`utils > renderAttributes > Should render attributes 2`] = `""`;

exports[`utils > renderAttributes > Should render attributes 3`] = `""`;

exports[`utils > renderAttributes > Should render attributes 4`] = `"name="batman""`;

exports[`utils > renderAttributes > Should render attributes 5`] = `"name="<div class='blu'>blu</div>""`;

exports[`utils > renderAttributes > Should render attributes 6`] = `"name="batman" user="bruce" active="true" id="1""`;

exports[`utils > renderAttributes > Should render attributes 7`] = `"empty-string="""`;

exports[`utils > renderAttributes > Should render attributes 8`] = `"name="0""`;

exports[`utils > renderAttributes > Should render attributes 9`] = `""`;

exports[`utils > renderAttributes > Should render attributes 10`] = `"name="""`;

exports[`utils > renderAttributes > Should render attributes 11`] = `""`;

exports[`utils > renderAttributes > Should render attributes 12`] = `""`;

exports[`utils > renderAttributes > Should render attributes 13`] = `""`;

exports[`utils > renderProperties > Should not render attributes as properties 1`] = `""`;

exports[`utils > renderProperties > Should not render attributes as properties 2`] = `""`;

exports[`utils > renderProperties > Should not render attributes as properties 3`] = `""`;

exports[`utils > renderProperties > Should not render attributes as properties 4`] = `
"
  "
`;

exports[`utils > renderProperties > Should render properties, case id selector 1`] = `
"
  document.querySelector('div').object={};document.querySelector('div').array=[]"
`;

exports[`utils > renderProperties > Should render properties, case id selector 2`] = `
"
  document.querySelector('div').object={};document.querySelector('div').array=[];document.querySelector('div').fn=() => {\\n    }"
`;
